GPU: Continue porting CUDA features to HIP: Main class files are now fully aligned.

davidrohr opened the pull request:
Doesn't mean that all features are already working in HIP, but the normal tracking remains working, and we can now start to generate the HIP files from the CUDA files.

davidrohr committed:
GPU: Port RTC source file and CMake part to HIP (not yet working)

davidrohr committed:
GPU: OpenMP should be used only when available

davidrohr committed:
GPU: GenRTC files can be compiled with gcc (to be able to use OpenMP in RedHat, which for some reason doesn't work with HIP)

davidrohr committed:
GPU: Extend CUDA compute target detection

davidrohr committed:
GPU: Align CUDA and HIP main class files

alibuild commented:
Error while checking build/O2/fullCI for 0c9a9f7edf8213dab99c2d9bc7ab1c8120591dc9 at 2024-03-14 12:47:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
clang-16: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12864/0c9a9f7edf8213dab99c2d9bc7ab1c8120591dc9/build_O2_fullCI/pretty.html).

davidrohr committed:
GPU: Fix deinitialization order

Was wrong for a long time, but not spotted since return value was not checked before

davidrohr committed:
GPU: ROCm HIP seems to fail linking with --fopenmp, as it saerches for missing -lomp on the EPNs.
